MericCelik/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,291 @@
+     * @param node1 : first node to consider
+     * @param node2 : other node to consider
+     * @return lowest common ancestor node
+     * */
+
+    public BinaryTreeNode<T> lowestCommonAncestor(BinaryTreeNode<T> node1, BinaryTreeNode<T> node2){
+        if(node1==null&&node2==null){
+            System.out.println("There are no first and second node!");
+            return null;
+        }
+        if(node1==null){
+            System.out.println("There is no first node!");
+            return null;
+        }
+        if(node2==null){
+            System.out.println("There is no second node!");
+            return null;
+        }
+        if(this.getRoot()==null){
+            System.out.println("The tree is empty!");
+            return null;
+        }
+
+        if(node1.equals(this.getRoot())&&node2.equals(this.getRoot())){
+            System.out.println("Both of them are root!");
+            return null;
+        }
+
+        ArrayList<BinaryTreeNode<T>> ancestors1 = findAncestors(node1);
+        int ancestor_count1 = ancestors1.size();
+
+        ArrayList<BinaryTreeNode<T>> ancestors2 = findAncestors(node2);
+        int ancestor_count2 = ancestors2.size();
+
+        if(ancestor_count1==0 && ancestor_count2==0){
+            System.out.println("Both nodes have no ancestors!");
+            return null;
+        }
+
+        if(node1.equals(this.getRoot())){
+            if(ancestors2.contains(node1))
+                return this.getRoot();
+            else{
+                System.out.println("Second node is not in the tree!");
+                return null;
+            }
+        }
+        if(node2.equals(this.getRoot())){
+            if(ancestors1.contains(node2))
+                return this.getRoot();
+            else{
+                System.out.println("First node is not in the tree!");
+            }
+        }
+
+
+        if(!this.getRoot().equals(ancestors1.get(ancestor_count1-1))){
+            System.out.println("First node is not in the tree!");
+            return null;
+        }
+        if(!this.getRoot().equals(ancestors2.get(ancestor_count2-1))){
+            System.out.println("Second node is not in the tree!");
+            return null;
+        }
+
+        if(ancestors2.contains(node1))
+            return node1;
+        if(ancestors1.contains(node2))
+            return node2;
+
+
+        int index1=0, index2=0;
+        if(ancestor_count1>ancestor_count2)
+            index1 = ancestor_count1-ancestor_count2;
+        else if(ancestor_count2>ancestor_count1)
+            index2 = ancestor_count2-ancestor_count1;
+        while(!ancestors1.get(index1).equals(ancestors2.get(index2))){
+            index1+=1;
+            index2+=1;
+        }
+        return ancestors1.get(index1);
+    }
+
+
+    public static void main(String[] args){
+
+        //Same example with the pdf's
+        BinaryTreeNode<Integer> six = new BinaryTreeNode<>(6);
```

Comment:
optional: This way of constructing the tree is good. In interviews it is common to use in-order, post-order or pre-order descriptions using brackets "(" and ")" for writing and reading a tree structure. If you'd like to practice doing so, you can add two methods to your tree - one that will print the whole structure using the brackets, the other that will read the description and create the tree based on it.

Reply:
I did not add yet but I definitely practise this too! Thank you! :smile: